go-controller/pkg/libovsdbops/db_object_ids.go

```diff
@@ -0,0 +1,283 @@
+	externalIDs := map[string]string{
+		OwnerControllerKey.String(): objectIDs.ownerControllerName,
+		OwnerTypeKey.String():       string(objectIDs.idsType.ownerObjectType),
+	}
+	for _, key := range objectIDs.idsType.GetExternalIDKeys() {
+		if _, ok := objectIDs.objectIDs[key]; !ok && omitUnset {
+			// unset ids are only omitted when omitUnset is true
+			continue
+		}
+		externalIDs[key.String()] = objectIDs.objectIDs[key]
+	}
+	return externalIDs
+}
+
+// GetDbObjectIDsFromExternalIDs is used to parse object ExternalIDs, it sets DbObjectIDs.ownerControllerName based
+// on OwnerControllerKey key, and verifies OwnerControllerKey value matches given objectIDsType.
+// All the other ids (except for PrimaryIDKey) will be set to DbObjectIDs.objectIDs, even keys that are not a part of objectIDsType.
+func GetDbObjectIDsFromExternalIDs(objectIDsType ObjectIDsType, externalIDs map[string]string) (*DbObjectIDs, error) {
+	if externalIDs[OwnerTypeKey.String()] != string(objectIDsType.ownerObjectType) {
+		return nil, fmt.Errorf("expected ExternalID %s to equal %s, got %s",
+			OwnerTypeKey, string(objectIDsType.ownerObjectType), externalIDs[OwnerTypeKey.String()])
+	}
+	if externalIDs[OwnerControllerKey.String()] == "" {
+		return nil, fmt.Errorf("required ExternalID %s is empty", OwnerControllerKey)
+	}
+	objIDs := map[ExternalIDKey]string{}
+	for key, value := range externalIDs {
+		if key != OwnerTypeKey.String() && key != OwnerControllerKey.String() && key != PrimaryIDKey.String() {
+			objIDs[ExternalIDKey(key)] = value
+		}
+	}
+	return NewDbObjectIDs(objectIDsType, externalIDs[OwnerControllerKey.String()], objIDs), nil
+}
+
+// hasExternalIDs interface should only include types that use new ExternalIDs from DbObjectIDs.
+type hasExternalIDs interface {
+	GetExternalIDs() map[string]string
+}
+
+// GetNoOwnerPredicate should only be used on initial sync when switching to new ExternalIDs.
+// Otherwise, use GetPredicate with the specific OwnerControllerKey id.
+func GetNoOwnerPredicate[T hasExternalIDs]() func(item T) bool {
+	return func(item T) bool {
+		return item.GetExternalIDs()[OwnerControllerKey.String()] == ""
+	}
+}
+
+// GetPredicate returns a predicate to search for db obj of type T (a set of supported types is defined in hasExternalIDs).
+// Only non-empty ids will be matched (that always includes DbObjectIDs.OwnerTypeKey and DbObjectIDs.ownerControllerName),
+// but
+// Other IDs may be empty and will be ignored in the filtering, additional filter function f may be passed, or set to nil.
+func GetPredicate[nbdbT hasExternalIDs](dbIndex *DbObjectIDs, f func(item nbdbT) bool) func(item nbdbT) bool {
+	primaryID, err := dbIndex.getUniqueIDWithCheck()
+	var indexIDs map[string]string
+	if err == nil {
+		// when primary id is set, other ids are not required
+		indexIDs = map[string]string{PrimaryIDKey.String(): primaryID}
+	} else {
+		indexIDs = dbIndex.getExternalIDsWithoutPrimary(true)
```

Comment:
We probably could just use `GetExternalIDs` here if it just didn't set the primary id if not available